### ISSUE TITLE
[Bug] Llama 3.1/3.2 model in FC mode output continue past where it should stop

### ISSUE THREAD

**Issue by HuanzhiMao** [sgl-project/sglang#1750]
### Checklist

- [X] 1. I have searched related issues but cannot get the expected help.
- [X] 2. The bug has not been fixed in the latest version.
- [X] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [X] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [X] 5. Please use English, otherwise it will be closed.

### Describe the bug

When using Llama 3.1/3.2 models like `meta-llama/Llama-3.2-3B-Instruct` in their function calling mode, sglang would continue past where it should stop (eg, `assistant\n\n`). Using vllm as the backend will not produce this issue. 

For example, using the same prompt and temperature settings, etc. 
vllm outputs
```
{\"name\": \"calculate_triangle_area\", \"parameters\": {\"base\": \"10\", \"height\": \"5\"}}
```

while sglang would output 
```
"{\"name\": \"calculate_triangle_area\", \"parameters\": {\"base\": \"10\", \"height\": \"5\"}}assistant\n\ndef calculate_triangle_area(base, height, unit=\"units\"):\n    return 0.5 * base * height\n\nprint(calculate_triangle_area(10, 5))assistant\n\nIn this code, we define a function `calculate_triangle_area` that takes three parameters: `base`, `height`, and `unit`. The `unit` parameter has a default value of \"units\" if not provided. The function calculates the area of a triangle using the formula `0.5 * base * height` and returns the result.\n\nWhen we call the function with `base=10` and `height=5`, it calculates the area of a triangle with a base of 10 units and a height of 5 units, and prints the result."
```

This seems like a Llama-specific issue. I did not observe it when using Llama in the normal chat mode, nor other models like Gemma 2, xLAM, etc. 

### Reproduction

`python -m sglang.launch_server --model-path meta-llama/Llama-3.2-3B-Instruct --attention-backend flashinfer`

```
client = OpenAI(base_url=f"http://localhost/:1053/v1", api_key="EMPTY")    
formatted_prompt='<|begin_of_text|><|start_header_id|>system<|end_header_id|>\n\nEnvironment: ipython\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\n<|eot_id|><|start_header_id|>user<|end_header_id|>\n\nGiven the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.\n\nRespond in the format {"name": function name, "parameters": dictionary of argument name and its value}.Do not use variables.\n\n{\n    "name": "calculate_triangle_area",\n    "description": "Calculate the area of a triangle given its base and height. Note that the provided function is in Python 3 syntax.",\n    "parameters": {\n        "type": "dict",\n        "properties": {\n            "base": {\n                "type": "integer",\n                "description": "The base of the triangle."\n            },\n            "height": {\n                "type": "integer",\n                "description": "The height of the triangle."\n            },\n            "unit": {\n                "type": "string",\n                "description": "The unit of measure (defaults to \'units\' if not specified)"\n            }\n        },\n        "required": [\n            "base",\n            "height"\n        ]\n    }\n}\n\nFind the area of a triangle with a base of 10 units and height of 5 units.<|eot_id|><|start_header_id|>assistant<|end_header_id|>\n\n'    
api_response = client.completions.create(
    model="meta-llama/Llama-3.2-3B-Instruct",
    temperature=0.001,
    prompt=formatted_prompt,
    max_tokens=4096,
)
```

### Environment

```
2024-10-22 06:15:31.990529: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-10-22 06:15:32.007851: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-10-22 06:15:32.030825: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8463] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-10-22 06:15:32.037602: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-10-22 06:15:32.053158: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX512F AVX512_VNNI AVX512_BF16 AVX512_FP16 AVX_VNNI, in other operations, rebuild TensorFlow with the appropriate compiler flags.
/usr/lib/python3/dist-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.17.3 and <1.25.0 is required for this version of SciPy (detected version 1.26.4
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"
Python: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0]
CUDA available: True
GPU 0: NVIDIA H100 PCIe
GPU 0 Compute Capability: 9.0
CUDA_HOME: /usr
NVCC: Cuda compilation tools, release 12.4, V12.4.131
CUDA Driver Version: 550.90.12
PyTorch: 2.4.0+cu121
sglang: 0.3.4.post1
flashinfer: 0.1.6+cu124torch2.4
triton: 3.0.0
transformers: 4.45.2
requests: 2.32.3
tqdm: 4.66.5
numpy: 1.26.4
aiohttp: 3.10.10
fastapi: 0.115.2
hf_transfer: 0.1.8
huggingface_hub: 0.26.1
interegular: 0.3.3
packaging: 21.3
PIL: 10.4.0
psutil: 5.9.0
pydantic: 2.9.2
uvicorn: 0.32.0
uvloop: 0.21.0
zmq: 22.3.0
vllm: 0.6.3.post1
multipart: 0.0.12
openai: 1.46.0
anthropic: 0.31.1
NVIDIA Topology: 
        GPU0    NIC0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      PHB     0-25    0               N/A
NIC0    PHB      X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0


Hypervisor vendor: KVM
ulimit soft: 1048576
```

**Comment by merrymercy** [comment]
This should be fixed by #1766.
Llama 3.2 will generate <|eom_id|> for tool use and the engine should stop at <|eom_id|> as well.

**Comment by SinanAkkoyun** [comment]
Hi, are special tokens enabled by default? If not, how can I enable them?
